### ISSUE TITLE
Topics for advanced react workshop

### ISSUE THREAD

**Issue by akiran** [ReactBangalore/react-advanced-workshop#1]
This thread is to discuss the agenda for advanced react workshop

I am thinking of these topics

* Higher order components (Nitish)
* Render props  (Nitish)
* refs (Kiran)
* Server side rendering (Kiran)
* Performance Optimizations (Darpan)
  * Understanding shouldComponentUpdate
  * Avoiding wasteful rerenders using PureComponent
  * Immutable data structures
  * Profiling Components with the Chrome Performance Tab
* Build tooling (Darpan)

Let me know, if we need to add or remove any topics from this list. 

@kakadiadarpan @sunnykgupta @nitishkr88 @kamleshchandnani  
Also let me know which topics you would like to present, so that we can started creating slides and exercises.

**Comment by nitishkr88** [comment]
Would we have time to cover **static type checking with Flow**?

I can take HOCs and Render Props as they are related.

**Comment by akiran** [comment]
static type checking with Flow is specific to react. So we can skip it for this workshop.

I didn't add context because its API is changing.
Should we cover portals and error boundaries in this workshop ?

**Comment by kakadiadarpan** [comment]
@akiran in my opinion we can remove findDOMNode as it doesn't get used by 90-95% of the developers. What do you say?

**Comment by akiran** [comment]
@kakadiadarpan,  Yes I agree.

**Comment by kakadiadarpan** [comment]
How about the following topics:

- Error boundaries
- Testing react components
- Performance Optimizations
  - Understanding shouldComponentUpdate
  - Avoiding wasteful rerenders using PureComponent
  - Immutable data structures
  - Profiling Components with the Chrome Performance Tab
- Build tooling

**Comment by kakadiadarpan** [comment]
I can take up **performance optimization** and **build tooling** if we decide to include it.

**Comment by akiran** [comment]
@kakadiadarpan Yes, I added those topics. You can start creating slides and exercises for them.

@nitishkr88 Yes, you can take render Props and HOC's. 
 
Feel free to open a PR with you content, so that we can review each others slides and examples and try to get good flow for workshop

**Comment by kamleshchandnani** [comment]
@iamhimadri are you interested to take up Testing react components?

**Comment by iamhimadri** [comment]
Sure @kamleshchandnani and @akiran I can take Testing React Component topic.

**Comment by akiran** [comment]
@kakadiadarpan I removed build tools from this workshop. We can have separate workshop for topics independent of React like build tools, code splitting, flow etc

@kamleshchandnani @iamhimadri I added your topics to the list

**Comment by kakadiadarpan** [comment]
@akiran sure! That makes sense actually 😄

**Comment by iamhimadri** [comment]
@kakadiadarpan @akiran @kamleshchandnani @nitishkr88 @souvikbasu
how are we following the folder structure for PR?
```
exercise/
     performance-optimizations/
     testing-react-component/
solution/
     performance-optimizations/
     testing-react-component/
```
OR
```
performance-optimizations/
    exercise/
    solution/
testing-react-component/
    exercise/
    solution/
```

**Comment by kakadiadarpan** [comment]
@iamhimadri 

```
exercise/
     performance-optimizations/
     testing-react-component/
solution/
     performance-optimizations/
     testing-react-component/
```

Let's keep the solutions separate from exercises.

**Comment by akiran** [comment]
@kakadiadarpan @akiran @kamleshchandnani @nitishkr88 @souvikbasu

excercises and solutions folder should be created with create react-react-app

you can put you code inside the source folder of exercises and solutions
